src/Elastic.Apm/Agent.cs

```diff
@@ -162,16 +171,31 @@ private static bool CheckAndAddFilter(Action<PayloadSenderV2> action)
 
 		public static void Setup(AgentComponents agentComponents)
 		{
-			if (LazyApmAgent.IsValueCreated)
-				agentComponents?.Logger?.Error()
-					?.Log("The singleton APM agent has" +
-						" already been instantiated and can no longer be configured. Reusing existing instance");
-
-			agentComponents?.Logger?.Trace()
-				?.Log("Initialization - Agent.Setup called. Callstack: {callstack}", new StackTrace().ToString());
-
-			Components = agentComponents;
-			_isConfigured = true;
+			lock (InitializationLock)
+			{
+				if (_isConfigured)
+				{
+					Components?.Logger?.Error()
+						?.Log("The singleton APM agent has" +
+							" already been instantiated and can no longer be configured. Reusing existing instance. "
```

Comment:
I just realized that technically this error message will be wrong/misleading when you call `Agent.Setup` **twice**.
In that case, no agent is actually instantiated at all, but the log message says so.
That was probably the reason for the existing code to check on the `LazyApmAgent.IsValueCreated` value.

Maybe bringing these two code parts (`LazyApmAgent` and `Setup`) together would make our code more readable.

Reply:
👍  this is a good point.

I'm not sure the `_isConfigured` variable is needed. I removed it and now it only relies on `LazyApmAgent`.